Announcing HogQL (#5469)
PostHog/posthog.com: contents/blog/introducing-hogql.md

## HogQL filters
![HogQL filter](../images/blog/array/hog_filters.gif)

Oh, look. You can use HogQL in filters too. Useful for _filtering_ by multiple properties!

## SQL insights
![PostHog SQL insights](../images/features/hogql/sql.gif)

If you're familiar with SQL, the easiest and most powerful way to leverage it within PostHog is via the new SQL insight type. This gives you direct SQL access to your data in PostHog, so you can create custom tables, visualization and more. 

SQL insights are especially useful for when you need to do exotic or novel analysis that isn't immediately possible with existing insight types. For example, while you can use a retention insight to discover which product features keep users coming back for more, you could use an SQL insight to find which users keep coming back and identify outliers. 

We're incredibly excited to hear your feedback and see what's possible with SQL access in PostHog, and believe it will give engineers a unique edge and better information than they'd be able to get with other product analytics tools.

> To give you a sense of what's possible with SQL, here's an example query we use to summarize data from our from [our 404 page](/hedgehogsgalore) and display which countries prefer their pizzas with pineapple on them.

Reviewer: Sadly we have a new 404 page without the survey. Thus this link is broken, but not in the right way.

Author: ```suggestion
To give you a sense of what's possible with SQL, here's an example query we use to summarize data from a survey we ran to determine which countries prefer pineapple on a pizza.
```

## Custom insights
![custom insights](../images/blog/array/custom_insights.gif)

Finally, we're introducing custom insights, which enable you to create a very wide range of insights and visualizations. 

Custom insights basically turn any list of events or persons - almost every table in PostHog, basically - into an insight which can be added to a dashboard. 

You can add tables directly by hitting the new 'Open as a new insight' button (above) to reveal the source behind the table, and editing it to create new views. You can also also use the 'View source' button on existing insights to accomplish the same. 

## The history and future of HogQL

We originally started thinking about HogQL back in January, while thinking about [a concept for universal search within PostHog](https://github.com/PostHog/posthog/issues/7963). Over time, that plan evolved into creating [new ways for users to explore data](https://github.com/PostHog/meta/issues/86) via direct queries. We wanted users to be able to run formulas powered by HogQL in what was then called the Live Events view, but is now the Event Explorer. 

> However, while this work was underway we realized we could potentially take it a step further and [build full SQL support](https://github.com/PostHog/meta/issues/86) directly into PostHog as a new insight type. You'd simply write a new SQL query in PostHog, press 'Run' and get a table with your desired results. 

Reviewer: ```suggestion
However, while this work was underway we realized we could potentially take it a step further and [build full SQL support](https://github.com/PostHog/meta/issues/81) directly into PostHog as a new insight type. You'd simply write a new SQL query in PostHog, press 'Run' and get a table with your desired results. 
```

---
date: 2023-05-08
title: "Introducing SQL and HogQL access for PostHog"
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
category: PostHog news
author:
  - joe-martin
tags:
  - Product updates
  - Release notes
featuredImage: ../images/blog/hog_ql.png
featuredImageType: full
---

Today, we're excited to announce a major new feature which we're making available to all users as a public beta on PostHog Cloud - the ability to directly access data stored in PostHog via SQL.

We've added several new ways to use SQL in PostHog - you can use expressions to enhance insights, you can use SQL to filter a revamped events list, and you can even use full SQL queries to analyse data in any way you want. 

> We've even created [our own SQL dialect, called HogQL](/docs/product-analytics/hogql), which enables direct access to data without the insecurities that come from using ClickHouse SQL directly.

Reviewer: "without the insecurities" is irrelevant from an user's standpoint.

We built HogQL prevent users from accessing one another's data, as everything is in just one database for us. But the users don't care about that, so this "we fixed insecurities" clause doesn't tell them anything.

We could either say "we found a way to hide your data from other posthog users when making sql queries", but that sounds like table stakes. So instead I'd just remove this bit... 🤔

PostHog's existing insights will remain valuable for the vast majority of queries, but this gives them a massive upgrade, and the ability to create your own SQL queries provides an alternative when off-the-shelf insights don't fit the bill. 

Speaking of bills, these features are free for all users while in public beta - all we ask is that you [share any feedback with us](http://app.posthog.com/home#supportModal). 

## What is HogQL?
HogQL is basically our take on SQL. 

Less basically, it's a transition layer over ClickHouse SQL that we've created to empower a wide range of users, from technical product managers who need multi-property breakdowns, to engineers who need to dissect data in more nuanced ways. It's intuitive for teams who are familiar with SQL, but offers a few extra advantages and features.

Unique features to HogQL include simplified access to event and person properties, as well as automatically added joins when you query fields with data on a different table, such as `events.person.properties.$browser`.

> You can also use aggregations, such as `count`, `min`, `minIf`, `max`, `maxIf`, `sum`, `sumIf`, `avg`, `any`, and `anyIf`. 

Reviewer: If we're listing language features, then there are far more impressive things than a few standard SQL aggregations. E.g.

> You can also use subqueries, joins, table expressions, arrays, aggregations, lambdas and a whole host of other neat SQL features.

We recommend checking the docs for a more detailed [explanation of ClickHouse SQL functions which are supported in HogQL](/manual/hogql).

## HogQL breakdowns

> ![HogQL breakdowns](../images/blog/array/hog_breakdown.gif)

Reviewer 2: @joethreepwood Can we switch these gifs out for MP4s?

Author: was planning to, aye

Reviewer: Could we actually show breaking something down by HogQL, instead of just opening the popup? E.g. `concat(properties.$browser, ' ', properties.$browser_version)`. Same for filters. Perhaps `properties.$screen_width < properties.$screen_height` to show only "portrait mode" users?

Reviewer: One other cool use of hogql is coming with this PR: https://github.com/PostHog/posthog/pull/15429

You can now aggregate funnels on any hogql expression.

## SQL insights
![PostHog SQL insights](../images/features/hogql/sql.gif)

If you're familiar with SQL, the easiest and most powerful way to leverage it within PostHog is via the new SQL insight type. This gives you direct SQL access to your data in PostHog, so you can create custom tables, visualization and more. 

SQL insights are especially useful for when you need to do exotic or novel analysis that isn't immediately possible with existing insight types. For example, while you can use a retention insight to discover which product features keep users coming back for more, you could use an SQL insight to find which users keep coming back and identify outliers. 

We're incredibly excited to hear your feedback and see what's possible with SQL access in PostHog, and believe it will give engineers a unique edge and better information than they'd be able to get with other product analytics tools.

To give you a sense of what's possible with SQL, here's an example query we use to summarize data from our from [our 404 page](/hedgehogsgalore) and display which countries prefer their pizzas with pineapple on them.

```

>    select properties.$geoip_country_name,
>           if(countIf(not properties.does_pineapple_go_on_pizza) > countIf(properties.does_pineapple_go_on_pizza),
>             '🍅 Does not belong on pizza',
>             if(countIf(not properties.does_pineapple_go_on_pizza) = countIf(properties.does_pineapple_go_on_pizza),
>               '🥦 It is a tie',
>               '🍍 Belongs on pizza'
>           )) as Result,
>           concat(
>              repeat('🍍', countIf(properties.does_pineapple_go_on_pizza)),
>              repeat('🍅', countIf(not properties.does_pineapple_go_on_pizza))
>           ) as Answers,
>           count() as `Number of answers`,
>           concat(
>             toString(round(countIf(properties.does_pineapple_go_on_pizza) / count() * 1000) / 10), 
>             '%'
>           ) as `Percentage pineapple`
>      from events
>     where event = 'pineapple_on_pizza_survey'
>  group by properties.$geoip_country_name
>  order by count() desc
>     limit 100

Reviewer: Can we swap it for this version, where the `if` looks a bit cleaner:

```suggestion
   select properties.$geoip_country_name,
          countIf(not properties.does_pineapple_go_on_pizza) > countIf(properties.does_pineapple_go_on_pizza)
            ? '🍅 Does not belong on pizza'
            : countIf(not properties.does_pineapple_go_on_pizza) = countIf(properties.does_pineapple_go_on_pizza)
              ? '🥦 It is a tie'
              : '🍍 Belongs on pizza'
          as Result,
          concat(
             repeat('🍍', countIf(properties.does_pineapple_go_on_pizza)),
             repeat('🍅', countIf(not properties.does_pineapple_go_on_pizza))
          ) as Answers,
          count() as `Number of answers`,
          concat(
            toString(round(countIf(properties.does_pineapple_go_on_pizza) / count() * 1000) / 10), '%'
          ) as `Percentage pineapple`
     from events
    where event = 'pineapple_on_pizza_survey'
 group by properties.$geoip_country_name
 order by count() desc
    limit 100
```

The event explorer enables you to view and edit the JSON source behind the table, so you can create tables in PostHog and add them to dashboards. You can also use HogQL expressions (e.g. `properties.$screen_width * $properties.screen_height`) and aggregations (e.g. `sum(properties.price)`) as columns.

This is helpful not just for generating tables which summarize person and event information according to your needs, but also for exploring the data in new and totally custom ways. You could, for example, use HogQL to run currency conversions and normalize data into a single currency in its own column - or even extrapolate and project revenue and usage figures into the future.

## Custom insights
![custom insights](../images/blog/array/custom_insights.gif)

Finally, we're introducing custom insights, which enable you to create a very wide range of insights and visualizations. 

Custom insights basically turn any list of events or persons - almost every table in PostHog, basically - into an insight which can be added to a dashboard. 

> You can add tables directly by hitting the new 'Open as a new insight' button (above) to reveal the source behind the table, and editing it to create new views. You can also also use the 'Edit source' button on existing insights to accomplish the same. 

Reviewer: ```suggestion
You can add tables directly by hitting the new 'Open as a new insight' button (above) to reveal the source behind the table, and editing it to create new views. You can also also use the 'View source' button on existing insights to accomplish the same. 
```

I think "edit source" makes some more sense, but the button is currently called "View source", so 🤷

## The history and future of HogQL

We originally started thinking about HogQL back in January, while thinking about [a concept for universal search within PostHog](https://github.com/PostHog/posthog/issues/7963). Over time, that plan evolved into creating [new ways for users to explore data](https://github.com/PostHog/meta/issues/86) via direct queries. We wanted users to be able to run formulas powered by HogQL in what was then called the Live Events view, but is now the Event Explorer. 

> However, while this work was underway we realized we could potentially take it a step further and build full SQL support directly into PostHog as a new insight type. You'd simply write a new SQL query in PostHog, press 'Run' and get a table with your desired results. 

Reviewer: ```suggestion
However, while this work was underway we realized we could potentially take it a step further and [build full SQL support](https://github.com/PostHog/meta/issues/86) directly into PostHog as a new insight type. You'd simply write a new SQL query in PostHog, press 'Run' and get a table with your desired results. 
```

You can also use subqueries, joins, table expressions, arrays, aggregations, lambdas and a whole host of other neat SQL features, including aggregations.

Whether you're a technical product manager who needs multi-property breakdowns, or an engineer wishing to dissect data in more nuanced ways, HogQL unlocks a whole new world of insights to PostHog users.

We recommend checking the docs for a more detailed [explanation of ClickHouse SQL functions which are supported in HogQL](/manual/hogql).

## HogQL breakdowns

![HogQL breakdowns](../images/blog/array/hog_breakdown.mp4)

<Caption>Using a `properties.$browser_version` HogQL expression to breakdown results</Caption>

A frequent request from teams using PostHog has been the ability to breakdown insight results across multiple properties. To accommodate this, we've added the ability to use HogQL expressions within a breakdown. 

In other words, you can now break down results however you like – including by multiple properties at once. Want to break down new signups by both pricing tier _and_ overall usage? Now you can.

> > **Ready to get started?** Read our tutorial on [how to use HogQL for running advanced breakdowns](docs/tutorials/hogql-breakdowns)

Reviewer 3: ```suggestion
> **Ready to get started?** Read our tutorial on [how to use HogQL for running advanced breakdowns](/tutorials/hogql-breakdowns)
```

## HogQL filters

![HogQL filter](../images/blog/array/hog_filters.mp4)

<Caption>Using a `properties.$screen_width < properties.$screen_height` HogQL expression as a filter</Caption>

Oh, look. You can use HogQL in filters too. Useful for _filtering_ by multiple properties!

> 

Reviewer 3: ```suggestion

> **Ready to get started?** Read our tutorial on [how to use HogQL for advanced time and date filters](/tutorials/hogql-date-time-filters)

```

> ## HogQL aggregations

Reviewer 3: I'm working on [this tutorial](https://github.com/PostHog/posthog.com/issues/6075) (almost done, but not a blocker), I've called it HogQL series. Do we think that is a better name? I'm neutral, can switch mine to aggregations.

Reviewer 2: I have no opinion... thoughts @mariusandra?

> ## HogQL aggregations

Reviewer 3: ```suggestion
## HogQL aggregations

```

![HogQL aggregations](../images/blog/array/hog_aggregate.mp4)
<Caption>Using a `properties.$session_id` HogQL expression to aggregate by unique sessions</Caption>

_And_ you can use HogQL to aggregate results in a funnel too. Is there anything HogQL can't do? Probably, yes, but we'll enjoy finding out for sure. 

> ## SQL insights

Reviewer 3: Are we sure this one is going to be out for the public beta?

Reviewer 2: @mariusandra can you confirm?

Reviewer: it's out

> ## SQL insights

Reviewer 3: ```suggestion
## SQL insights

```

![PostHog SQL insights](../images/features/hogql/pinepple_sql.gif)
<Caption>Using an SQL insight to run the query below</Caption>

> If you're familiar with SQL, the easiest and most powerful way to leverage it within PostHog is via the new SQL insight type. This gives you direct SQL access to your data in PostHog, so you can create custom tables, visualization and more. 

Reviewer 3: ```suggestion
If you're familiar with SQL, the most powerful way to leverage it within PostHog is via the new SQL insight type. This gives you direct SQL access to your data in PostHog, so you can create custom tables, visualizations, and more. 
```
The other ways are easier!

SQL insights are especially useful for when you need to do exotic or novel analysis that isn't immediately possible with existing insight types. For example, while you can use a retention insight to discover which product features keep users coming back for more, you could use an SQL insight to find which users keep coming back and identify outliers. 

> We can't wait to hear your feedback, and see what's possible with SQL access in PostHog. We're confident it will give engineers a unique edge, and better information, than other analytics tools.

Reviewer 3: ```suggestion
We can't wait to hear [your feedback](http://app.posthog.com/home#supportModal), and see what's possible with SQL access in PostHog. We're confident it will give engineers a unique edge, and better information, than other analytics tools.
```

To give you a sense of what's possible with SQL, here's an example query we use to summarize data from a survey we ran to determine which countries prefer pineapple on a pizza.

```
   select properties.$geoip_country_name,
          countIf(not properties.does_pineapple_go_on_pizza) > countIf(properties.does_pineapple_go_on_pizza)
            ? '🍅 Does not belong on pizza'
            : countIf(not properties.does_pineapple_go_on_pizza) = countIf(properties.does_pineapple_go_on_pizza)
              ? '🥦 It is a tie'
              : '🍍 Belongs on pizza'
          as Result,
          concat(
             repeat('🍍', countIf(properties.does_pineapple_go_on_pizza)),
             repeat('🍅', countIf(not properties.does_pineapple_go_on_pizza))
          ) as Answers,
          count() as `Number of answers`,
          concat(
            toString(round(countIf(properties.does_pineapple_go_on_pizza) / count() * 1000) / 10), '%'
          ) as `Percentage pineapple`
     from events
    where event = 'pineapple_on_pizza_survey'
 group by properties.$geoip_country_name
 order by count() desc
    limit 100
```

> ## The PostHog event explorer 

Reviewer 3: ```suggestion
## The PostHog event explorer 

```

We haven't introduced HogQL in isolation – we've also bought it into existing parts of PostHog, supercharging them to deliver new functionality. 

> The Event Explorer (previously the Live Events tab) is a great example of this. What was previously a stream of incoming events now offers wholly new ways to interact with your data and filter it into custom tables. 

Reviewer 3: ```suggestion
The "Event Explorer" (previously the "Live Events" tab) is a great example of this. What was previously a stream of incoming events now offers wholly new ways to interact with your data and filter it into custom tables. 
```

> The event explorer enables you to view and edit the JSON source behind the table, so you can create tables in PostHog and add them to dashboards. You can also use HogQL expressions (e.g. `properties.$screen_width * $properties.screen_height`) and aggregations (e.g. `sum(properties.price)`) as columns.

Reviewer 3: ```suggestion
The event explorer enables you to view and edit the JSON schema behind the table, so you can create tables in PostHog and add them to dashboards. You can also use HogQL expressions (e.g. `properties.$screen_width * $properties.screen_height`) and aggregations (e.g. `sum(properties.price)`) as columns.
```

This is helpful not just for generating tables that summarize person and event information according to your needs, but also for exploring the data in new and totally custom ways. 

You could, for example, use HogQL to run currency conversions and normalize data into a single currency in its own column – or even extrapolate and project revenue and usage figures into the future.

> ## Custom insights

Reviewer 3: ```suggestion
## Custom insights

```

Reviewer 3: Also, is this not just saving event explorer views?

Reviewer 3: I was expecting this to be the "insight source" schema editor, rather than just saving from event explorer.

Reviewer 2: Good question. I was just writing a twitter thread for this and chose to focus on the 'view source' feature as it seemed more interesting (to me).

@mariusandra sorry to bug you again, but your input might be useful here

Reviewer: We're also slightly struggling with how to present this. The tldr is that thanks to "data exploration", everything is now represented with a JSON object that can be 1) put on a dashboard, 2) explored as an insight, etc. This includes data tables like the event explorer. You can now save those as insights, however you can also view/edit the source of traditional insights. I'm not sure what's the best way to serve this.

---

> date: 2023-06-06

Reviewer 2: ```suggestion
date: 2023-06-07
```

> title: "Introducing HogQL: Direct SQL access for PostHog"

Reviewer 2: Should we mention it's a public beta in the headline?

    limit 100
```

## The PostHog event explorer 

We haven't introduced HogQL in isolation – we've also bought it into existing parts of PostHog, supercharging them to deliver new functionality. 

The "Event Explorer" (previously the "Live Events" tab) is a great example of this. What was previously a stream of incoming events now offers wholly new ways to interact with your data and filter it into custom tables. 

The event explorer enables you to view and edit the JSON schema behind the table, so you can create tables in PostHog and add them to dashboards. You can also use HogQL expressions (e.g. `properties.$screen_width * $properties.screen_height`) and aggregations (e.g. `sum(properties.price)`) as columns.

This is helpful not just for generating tables that summarize person and event information according to your needs, but also for exploring the data in new and totally custom ways. 

You could, for example, use HogQL to run currency conversions and normalize data into a single currency in its own column – or even extrapolate and project revenue and usage figures into the future.

## Custom insights
![custom insights](../images/blog/array/custom_insights.gif)

> Finally, we're introducing custom insights, which enable you to create a very wide range of insights and visualizations. 
> 
> Custom insights basically turn any list of events or persons - almost every table in PostHog, basically - into an insight that can be added to a dashboard. 
> 
> You can add tables directly by hitting the new 'Open as a new insight' button (above) to reveal the source behind the table, and editing it to create new views. You can also use the 'View source' button on existing insights to accomplish the same. 

Reviewer 3: ```suggestion
Finally, we're introducing custom insights. All of our insights have been reworked as JSON objects which can fully customized by you.

Now any list of events or persons, almost every table in PostHog, can be customized with the data sources, filters, and visualizations you want.

You can create or edit custom insights in two ways:

1.  You can add tables from the event explorer directly by hitting the new 'Open as a new insight' button (above) to reveal the source behind the table, and editing it to create new views. 

2. You can also use the 'View source' button on existing insights to accomplish the same. 
```
Based on [Marius' comment](https://github.com/PostHog/posthog.com/pull/5469/files#r1220033857), I think this clears things up.

Reviewer 2: Legend, thank you.

---

> date: 2023-06-07

Reviewer 2: ```suggestion
date: 2023-06-08
```

title: "Introducing HogQL: Direct SQL access for PostHog"
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
category: PostHog news
author:
  - joe-martin
  - andy-vandervell
tags:
  - Product updates
  - Release notes
featuredImage: ../images/blog/hog_ql.png
featuredImageType: full
---

Today, we're releasing a major new feature as a public beta: the ability to directly query your PostHog data using SQL. We call this [HogQL](/docs/product-analytics/hogql) because... hedgehogs love SQL, probably?

You can use HogQL expressions to, among other things, enhance insights, filter event lists, and (most importantly) write full queries to analyze data in any way you want. 

> PostHog's existing insights are already incredibly powerful. HogQL is PostHog insights, but Super Saiyan.

Reviewer 3: ```suggestion
PostHog's existing insights are already incredibly powerful. HogQL turns insights Super Saiyan.
```

HogQL is free for all users while in public beta – all we ask is that you [share your feedback with us](http://app.posthog.com/home#supportModal).

## What is HogQL?

> It's a transition layer over ClickHouse SQL. It's intuitive for anyone familiar with SQL, but it offers a few advantages and features.

Reviewer 4: ```suggestion
It's a translation layer over ClickHouse SQL. It's intuitive for anyone familiar with SQL, but it offers a few advantages and features.
```

translation not transition?

> It's a translation layer over ClickHouse SQL. It's intuitive for anyone familiar with SQL, but it offers a few advantages and features.

Reviewer 3: ```suggestion
It's a translation layer over ClickHouse SQL. It's intuitive for anyone familiar with SQL, but it offers a few advantages and customizations.
```
It offers more than a few features

See our [HogQL documentation](/docs/product-analytics/hogql) for a full [list of supported ClickHouse SQL functions](/docs/product-analytics/hogql#supported-clickhouse-functions).

> **Not a PostHog user?** [Get started for free](https://app.posthog.com/signup?utm_source=hogql-blog-top) – all users get 1 million events and 15k recordings free every month, **no card required.**

## What can you do with HogQL?

Too much to list in one blog post, but here are a few examples:

### HogQL breakdowns = multiple properties

Want to break down signups by both pricing tier _and_ overall usage? No problem. HogQL does that. With HogQL, you can add as many breakdown properties as you like. Go nuts.

![HogQL breakdowns](../images/blog/array/hog_breakdown.mp4)

> **🎓 Related tutorial:** [Using HogQL for advanced breakdowns](/tutorials/hogql-breakdowns)

### HogQL filters

Oh, look. You can use HogQL in filters too. Useful for _filtering_ by multiple properties!

![HogQL filter](../images/blog/array/hog_filters.mp4)

> **🎓 Related tutorial** [Using HogQL for advanced time and date filters](/tutorials/hogql-date-time-filters)

### HogQL aggregations

> _And_ you can use HogQL to aggregate results in a funnel too, like we do with `properties.$session_id` above. Is there anything HogQL can't do? Probably, yes, but we'll enjoy finding out for sure. 

Reviewer 2: ```suggestion
_And_ you can use HogQL to aggregate results in a funnel, too. Is there anything HogQL can't do? Probably, yes, but we'll enjoy finding out for sure. 
```

![HogQL aggregations](../images/blog/array/hog_aggregate.mp4)

### Custom SQL insights

> Of course, the most powerful way to leverage HogQL within PostHog is via the new SQL insight type. This gives you direct SQL access to your data in PostHog, so you can create custom tables, visualizations, and more. 

Reviewer 2: ```suggestion
Of course, the most powerful way to leverage HogQL within PostHog is via the new SQL insight type. This gives you direct SQL access to your data in PostHog, so you can create custom table insights that answer complex questions.
```

For example, while you can use a retention insight to discover which **features** keep users coming back for more, you could build an SQL insight to find which **users** keep coming back, and identify outliers. 

> We're confident SQL insights will unlock unique insights into how users use your products, and we can't wait to hear [your feedback](http://app.posthog.com/home#supportModal) how you use SQL access in PostHog. 

Reviewer 2: ```suggestion
We're confident SQL insights will unlock deep analysis into how users use your products, and we can't wait to hear [your feedback](http://app.posthog.com/home#supportModal) and see how you use SQL access in PostHog. 
```

          concat(
             repeat('🍍', countIf(properties.does_pineapple_go_on_pizza)),
             repeat('🍅', countIf(not properties.does_pineapple_go_on_pizza))
          ) as Answers,
          count() as `Number of answers`,
          concat(
            toString(round(countIf(properties.does_pineapple_go_on_pizza) / count() * 1000) / 10), '%'
          ) as `Percentage pineapple`
     from events
    where event = 'pineapple_on_pizza_survey'
 group by properties.$geoip_country_name
 order by count() desc
    limit 100
```

## Event explorer + HogQL = 🚀 

There. Is. Mooooorrrre.

As part of our work on HogQL, we've reworked all our insights as JSON objects, which you can customize. Just click the 'View source' button in the top right (see below), and tweak the code directly.

![view source](../images/blog/array/view-source.mp4)

You can also create custom table insights directly from the Events Explorer (previously "Live Events") and Person & Groups tabs. 

![custom insights](../images/blog/array/custom_insights.mp4)

Just hit that 'Open as a new insight' button (see above), tweak the JSON (if you want), and save your table to a dashboard or notebook... 

> [_What's that? Notebooks haven't launched, yet? But they're the future of everyth... oh, my bad!"_]. 🥸

Reviewer 3: ```suggestion
_What's that? Notebooks haven't launched, yet? But they're the future of everyth... oh, my bad!"_. 🥸
```

Moving on... you can also use HogQL expressions (e.g. `properties.$screen_width * $properties.screen_height`) and aggregations (e.g. `sum(properties.price)`) as columns. This is helpful not just for generating tables that summarize person and event information, but also for exploring the data in new and totally custom ways. 

You could, for example, use HogQL to run currency conversions and normalize data into a single currency in its own column, or even extrapolate and project revenue, and usage figures, into the future. 

> ## The history (and future) of HogQL

Reviewer 2: ```suggestion
## Why did we build HogQL and what's next?
```

It started with a small idea and just got bigger.

We originally started thinking about HogQL back in January, while thinking about [a concept for universal search within PostHog](https://github.com/PostHog/posthog/issues/7963). 

Over time, that plan evolved into creating [new ways for users to explore data](https://github.com/PostHog/meta/issues/86) via direct queries. We wanted to enable to run formulas in what was then called the Live Events view, but is now the Event Explorer. 

However, while this work was underway we realized we could potentially take it a step further and build full SQL support directly into PostHog as a new insight type... so we did! It was a short conversation. 

We're still actively developing all these ideas, as well as the implementation of HogQL within PostHog – seriously, check out this [massive mega issue](https://github.com/PostHog/meta/issues/81). This is, as the cliché goes, just the beginning. There's at least 19 more Super Saiyan forms to go...

> Got an opinion on what we should do next? Share it in [the PostHog Slack group](/slack), or via the [feedback modal in PostHog](https://app.posthog.com/home#supportModal=feedback%3A).

Reviewer 3: ```suggestion
Got an opinion on what we should do next? Share it via the [feedback modal in PostHog](https://app.posthog.com/home#supportModal=feedback%3A).
```
Alternative to Slack: GitHub and Twitter?

> **Not a PostHog user?** [Get started for free](https://app.posthog.com/signup?utm_source=hogql-blog-bottom) – all users get 1 million events and 15k recordings free every month, **no card required.**

> [^1]: 68.5% of 🇺🇸 residents think 🍍 belongs on 🍕. People in 🇪🇸 are the greatest 🍍 deniers at 10%. One person in Aruba 🇦🇼 voted, but we're pretty sure that was one of us during our [2023 all-company offsite](/blog/aruba-hackathon). We built some cool hackathon projects there, like our [dashboard template library](/blog/aruba-hackathon), and an open-source tool for [monitoring and managing ClickHouse clusters](https://github.com/PostHog/HouseWatch). You could say it was openly... sourcey. "_Hello, HR? Are bad puns a firing offense?_"

Reviewer 3: ```suggestion
[^1]: 68.5% of 🇺🇸 residents think 🍍 belongs on 🍕. People in 🇪🇸 are the greatest 🍍 deniers at 10%. One person in Aruba 🇦🇼 voted, but we're pretty sure that was one of us during our [2023 all-company offsite](/blog/aruba-hackathon). We built some cool hackathon projects there, like our [dashboard template library](/templates), and an open-source tool for [monitoring and managing ClickHouse clusters](https://github.com/PostHog/HouseWatch). You could say it was openly... sourcey. "_Hello, HR? Are bad puns a firing offense?_"
```

---
date: 2023-06-07
title: "Introducing HogQL: Direct SQL access for PostHog"
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
category: PostHog news
author:
  - joe-martin
tags:
  - Product updates
  - Release notes
featuredImage: ../images/blog/hog_ql.png
featuredImageType: full
---

Today, we're excited to announce a major new feature that's now available for all users as a public beta on PostHog Cloud - the ability to directly query data stored in PostHog via SQL. We call this [HogQL](/docs/product-analytics/hogql).

We've added several new ways to use HogQL in PostHog - you can use expressions to enhance insights, you can use HogQL to filter a revamped events list, and you can even use full queries to analyze data in any way you want. 

> 

Reviewer 5: ```suggestion

```

```
   select properties.$geoip_country_name,
          countIf(not properties.does_pineapple_go_on_pizza) > countIf(properties.does_pineapple_go_on_pizza)
            ? '🍅 Does not belong on pizza'
            : countIf(not properties.does_pineapple_go_on_pizza) = countIf(properties.does_pineapple_go_on_pizza)
              ? '🥦 It is a tie'
              : '🍍 Belongs on pizza'
          as Result,
          concat(
             repeat('🍍', countIf(properties.does_pineapple_go_on_pizza)),
             repeat('🍅', countIf(not properties.does_pineapple_go_on_pizza))
          ) as Answers,
          count() as `Number of answers`,
          concat(
            toString(round(countIf(properties.does_pineapple_go_on_pizza) / count() * 1000) / 10), '%'
          ) as `Percentage pineapple`
     from events
    where event = 'pineapple_on_pizza_survey'
 group by properties.$geoip_country_name
 order by count() desc
    limit 100
```

## The PostHog event explorer 

We haven't introduced HogQL in isolation – we've also bought it into existing parts of PostHog, supercharging them to deliver new functionality. 

> The "Event Explorer" (previously the "Live Events" tab) is a great example of this. What was previously a stream of incoming events now offers wholly new ways to interact with your data and filter it into custom tables. 

Reviewer 5: ```suggestion
The "[Event Explorer](https://app.posthog.com/events)" (previously the "Live Events" tab) is a great example of this. What was previously a stream of incoming events now offers wholly new ways to interact with your data and filter it into custom tables. 
```

---
date: 2023-03-02
title: "Introducing HogQL for PostHog"
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
category: PostHog news
author:
  - joe-martin
tags:
  - Product updates
  - Release notes
featuredImage: ../images/blog/hog_ql.png
featuredImageType: full
---

Today, we're excited to announce a major new feature which we're making available to all users as a public beta on PostHog Cloud. It's called HogQL and it's a new query language which makes it easier to interrogate your data directly.

> Not only that, but the introduction of HogQL brings with it other updates, including a new event explorer (previously the Live Events view), and the ability to use HogQL expressions within insights. We think that PostHog's existing insights will remain valuable for the vast majority of queries, but HogQL is an alternative when they just don't fit the bill.

Reviewer 6: > We think that PostHog's existing insights will remain valuable for the vast majority of queries, but HogQL is an alternative when they just don't fit the bill.

HogQL also makes existing insights much more powerful

> Speaking of bills, HogQL is free for all users while it's in public beta - all we ask is that you share any feedback you have with it either via the PostHog Slack, or by [filing an issue on our repo](https://github.com/PostHog/).

Reviewer 6: Given we grandfather in pricing, I don't think we are planning to charge for this separately for existing users

## What is HogQL?

HogQL is basically our take on SQL. 

Less basically, it's a transition layer over ClickHouse SQL that we've created to empower a wide range of users, from technical product managers who need multi-property breakdowns, to engineers who need to dissect data in more nuanced ways. It's intuitive for teams who are familiar with SQL, but offers a few extra advantages and features.

Unique features to HogQL include simplified access to event and person properties, as well as the ability to run subqueries and database joins. You can also use aggregations, such as `count`, `min`, `minIf`, `max`, `maxIf`, `sum`, `sumIf`, `avg`, `any`, and `anyIf`.

We recommend checking the docs for a more detailed [explanation of ClickHouse SQL functions which are supported in HogQL](/manual/hogql).

## HogQL expressions
![HogQL trends breakdown filter](../images/features/hogql/trends-breakdown.png)

> A frequent request from teams using PostHog has been the ability to breakdown insight results across multiple properties. To accommodate this, we've added the ability to use HogQL expressions within an insight breakdown - so you can breakdown results however you like. 

Reviewer 6: I don't think we have multiple breakdowns yet, just the ability to do more advanced breakdowns with HogQL

Author: @mariusandra Can you confirm? Disagrees with my notes from our chat.

Reviewer: It's kind of the same thing if you know what you're doing. HogQL lets you create a combined "super-propety", even if just an array/tuple (`tuple(distinct_id, properties.bla)`) or a combined string of normal properties (`concat(properties.a, properties.b)`), and break down by that. This effectively serves the same purpose as the "multi breakdown interface" we could build.

We can now build a nicer interface around it that actually lets the user enter multiple values, and put them into a `tuple()` ourselves behind the hood.

There are other problems with HogQL breakdown filters. The biggest one is the UX bug that you can just add them, but you can't edit them. You have to delete the filter you added and make a new one. There's also some weird capitalisation that sometimes happens, etc.

It all needs some love, but technically, if you know what you're doing, you can break down by multiple properties :).

Unique features to HogQL include simplified access to event and person properties, as well as the ability to run subqueries and database joins. You can also use aggregations, such as `count`, `min`, `minIf`, `max`, `maxIf`, `sum`, `sumIf`, `avg`, `any`, and `anyIf`.

We recommend checking the docs for a more detailed [explanation of ClickHouse SQL functions which are supported in HogQL](/manual/hogql).

## HogQL expressions
![HogQL trends breakdown filter](../images/features/hogql/trends-breakdown.png)

A frequent request from teams using PostHog has been the ability to breakdown insight results across multiple properties. To accommodate this, we've added the ability to use HogQL expressions within an insight breakdown - so you can breakdown results however you like. 

HogQL expressions in insights enables you to combines the speed of PostHog's existing insights with the flexibility of HogQL - so you can do things such as plotting the signup sources by email domain, or breaking down signups by both pricing tier and overall usage.

> ## The PostHog event explorer 

Reviewer: A section called "HogQL insights" should go between the two. We're finally launching it today for internal usage 🤞

In addition to introducing HogQL, we've also upgraded the event explorer (previously known as the Live events tab) to give you more ways to filter events and analyze your data.

> The new event explorer enables you to view source behind queries, then use HogQL to create custom views. Below, for example, you can see a query we wrote using HogQL which summarizes [votes from our 404 page](/hedgehogsgalore).

Reviewer: ```suggestion
The event explorer also lets you view and edit the JSON source behind the table. Below, for example, you can see a query we wrote using HogQL which summarizes [votes from our 404 page](/hedgehogsgalore).
```

---
date: 2023-03-02
title: "Introducing HogQL, SQL and Custom insights for PostHog"
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
category: PostHog news
author:
  - joe-martin
tags:
  - Product updates
  - Release notes
featuredImage: ../images/blog/hog_ql.png
featuredImageType: full
---

> Today, we're excited to announce a major new feature which we're making available to all users as a public beta on PostHog Cloud. It's called HogQL and it's a new query language which makes it easier to interrogate your data directly.

Reviewer: The intent of this line reads as if we are "launching a new query language". However what we're really launching is the ability for anyone to make SQL queries against data stored in PostHog. I think that fact should be the headline feature. 

To make those SQL queries, you can:
- Write actual SQL and save it as an insight
- Use the "event explorer" view with columns that you can customise with "SQL expressions" (aka "chunks of SQL")
- Use "SQL expressions" in existing insights.

Now, our dialect of SQL is called HogQL, and the terms "SQL" and "HogQL" are interchangeable in a technical context. However emphasising that it's a new language makes it feel like a hurdle that I need to overcome.

Thus, saying we're "launching SQL support" will directly communicate the essence of what this change brings. Saying that we're "launching HogQL" makes you just wonder "What is that now?".

I wonder if we should just rename the blog post to "Announcing SQL support"?

Author: Running in some changes to facilitate this.

Below you can see a query we wrote using HogQL which summarizes [votes from our 404 page](/hedgehogsgalore).

```
{
  "kind": "DataTableNode",
  "full": true,
  "source": {
    "kind": "EventsQuery",
    "select": [
      "properties.$geoip_country_name",
      "if(countIf(properties.does_pineapple_go_on_pizza == 'false') > countIf(properties.does_pineapple_go_on_pizza == 'true'), '🍅 Does not belong on pizza', if(countIf(properties.does_pineapple_go_on_pizza == 'false') == countIf(properties.does_pineapple_go_on_pizza == 'true'), '🥦 It is a tie' , '🍍 Belongs on pizza')) -- Result",
      "concat(repeat('🍍', countIf(properties.does_pineapple_go_on_pizza == 'true')),repeat('🍅', countIf(properties.does_pineapple_go_on_pizza == 'false'))) -- Answers",
      "count() -- Number of answers",
      "concat(toString(round(countIf(properties.does_pineapple_go_on_pizza == 'true') / count() * 1000) / 10), '%') -- Percentage pineapple"
    ],
    "after": "-365d",
    "limit": 100,
    "event": "pineapple_on_pizza_survey",
    "orderBy": [
      "-count() # Number of answers"
    ],
    "before": ""
  },
  "propertiesViaUrl": true
}

> ```

Reviewer: I think this example would be [better as SQL](https://app.posthog.com/insights/K84UWNwR).

Author: "Insight not found" ?
